### PR TITLE
Support for null in mapAsync and fromCompletionStage was added. #25475

### LINKED ISSUE #25475: Support for CompletionStage<Void> in streams
`CompletionStage<Void>` is quickly becoming the de facto standard for signalling completion of an operation with no return value. This doesn't integrate well into Akka Streams, since such `CompletionStage`'s will always be redeemed with `null`, and Akka Streams doesn't allow nulls. For example, it's very common to see this kind of API for committing transactions or messages:

```java
CompletionStage<Void> commit();
```

If you want to use this from Akka Streams `mapAsync`, you need to:

```java
.mapAsync(1, message ->
  message.commit().thenApply(n -> Done.getInstance())
);
```

It's even worse if you want to convert a completion signal into a source that completes when the completion stage redeems:

```java
Source.fromCompletionStage(
  message.commit().thenApply(n -> Done.getInstance())
).filter(d -> false);
```

As much as Akka doesn't like nulls, `CompletionStage` does like nulls, and I think good integration with `CompletionStage` should take this into account.

My thoughts are that `mapAsync` and `fromCompletionStage` should support nulls, converting `null` to no signal in the case of `mapAsync`, or to just the complete signal with no element in the case of `fromCompletionStage`. Another alternative is to offer `fromCompletionStageNullable` and `mapAsyncNullable`.

### PR DESCRIPTION
**Pull request by mszczygiel**
Resolves #25475

### CODE REVIEW

**Comment by lightbend-cla-validator** [comment]
Hi @mszczygiel,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

**Comment by raboof** [comment]
OK TO TEST

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala [hunk 1]
```diff
@@ -1284,11 +1281,8 @@ private[stream] object Collect {
               if (!hasBeenPulled(in)) tryPull(in)
               push(out, elem)
             } else buffer.enqueue(elem)
-          case other ⇒
-            val ex = other match {
-              case Failure(t)              ⇒ t
-              case Success(s) if s == null ⇒ ReactiveStreamsCompliance.elementMustNotBeNullException
-            }
+          case Success(null) ⇒
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: pull here?
```
if (!hasBeenPulled(in)) tryPull(in)
```

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowMapAsyncSpec.scala [hunk 2]
```diff
@@ -304,24 +304,14 @@ class FlowMapAsyncSpec extends StreamSpec {
...
-      val p = Source(List("a", "b", "c"))
-        .mapAsync(4)(elem ⇒ if (elem == "b") Future.successful(null) else Future.successful(elem))
-        .withAttributes(supervisionStrategy(resumingDecider))
-        .to(Sink.fromSubscriber(c)).run()
-      val sub = c.expectSubscription()
-      sub.request(10)
-      for (elem ← List("a", "c")) c.expectNext(elem)
-      c.expectComplete()
+      result.futureValue should ===(Seq("1", "3"))
     }
```
Comment (patriknw) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowMapAsyncSpec.scala]: perhaps add some more test or add more to this one, eg mapAsync(3) and then return 5 null in a row and verify that elements after that are still working

**Review by patriknw — COMMENTED** [review]
looking good, some question

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala [hunk 3]
```diff
@@ -1226,7 +1223,7 @@ private[stream] object Collect {
           val holder = buffer.dequeue()
           holder.elem match {
             case Success(elem) ⇒
-              push(out, elem)
+              if (elem != null) push(out, elem)
```
Comment (mszczygiel) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: Stage will not complete if null element is emmited last. `pushNextIfPossible()` should be added in case of `null`. I will add a test case for this

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala (hunk 1 above)
Comment (mszczygiel) [reply akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: Thanks for spotting it! Actually I think
```
if (isClosed(in) && todo == 0) completeStage()
else if (!hasBeenPulled(in)) tryPull(in)
``` 
would be correct. Will add testcase to cover stage completion when null is emitted last.

**Comment by mszczygiel** [comment]
@patriknw PR ready the next round. Please take a look

**Review by patriknw — APPROVED** [review]
LGTM, thanks

**Comment by mszczygiel** [comment]
Please let me know if you want me to rebase this PR

**Comment by patriknw** [comment]
@mszczygiel I think this is a nice feature so rebasing would be welcome and then we can merge it (I have already approved it once). Thanks for patience.

**Comment by mszczygiel** [comment]
@patriknw I've created new PR with the same changes rebased onto recent master - https://github.com/akka/akka/pull/26721

**Comment by patriknw** [comment]
Superseded by https://github.com/akka/akka/pull/26721
